### DISCUSSION TITLE
" Your first reaction should be to redesign your API in order to get rid of all circular references" - But how?

### DISCUSSION THREAD

**Discussion by lennartb-** [AutoFixture/AutoFixture#1333] [Q&A]
We're currently building up an API that contains a hierarchy very similar to a file system: There is a root node, with an arbitrary number of folders, which in turn can themselves can have any number of folders as well as files (there is only one type of file):

```
public class Root
{
    public string Id {get; set;}
    public IEnumerable<Folder> Folders {get; set;}
}
```

```
public class Folder
{
    public string Id {get; set;}

    // This information comes from a DB
    public string ParentId {get; set;}

    public IEnumerable<File> Files {get; set;}
    public IEnumerable<Folder> Folders {get; set;}
}
```

```
public class File
{
    public string Id {get; set;}

    // This information comes from a DB
    public string ParentFolderId {get; set;}
}
```
There is currently no inheritance or interfaces involved, but since we have full control over this API, we can change this at will.

When trying to build a fixture for this, AF rightfully complains about the circular reference in `Folder`, which can be solved with the `OmitOnRecursionBehavior`.  But the initial error message, which contains the hint I mentioned in the title, stuck with me, and I have a gut feeling, that this hierarchy could be solved better. But how? The information which Files and Folders are in a Folder comes from a DB, which we map with EF to objects, and then manually build the hierarchy by putting the child objects into the respective lists of their parents based on a `ParentFolderId` we have.

The only similar, professionally designed, structure I know of is the `DirectoryInfo` and `FileInfo` of .NETs API, which use a `DirectoryInfo? Parent` for upwards references, and methods (`EnumerateDirectories()`...) instead of properties for child objects. Since we're heavily using MVVM and want to display this hierarchy in a TreeView or something, properties would make data-binding a lot easier than methods.

Is there a better way to design this? I know this doesn't directly relate to AF, but it made me question our design choices.

**Comment by aivascu** [comment]
@lennartb- this is my personal opinion, but a good starting point toward improving your API is to follow the OOP and SOLID guidelines. It helps to use TDD while developing new code. The tests usually guide you toward a better design since the flaws are more visible from the start. If you're working with a framework, try to follow as close as possible the guidelines for that framework. I know it sounds obvious, but most code doesn't follow these rules, and suffers because of that.

One of the things I think you could improve is encapsulation. IMO your properties should not have public setters and should restrict construction to allow only valid objects to be created.
`DirectoryInfo` is a good example of encapsulation. For example, it restricts construction by implementing a constructor that only accepts a valid path. Also adding sub-directories is allowed only via the `CreateSubdirectory()` method. It could have exposed a `List<DirectoryInfo> Subdirectories { get; set; }`, however that would open the door for a lot of issues. Imagine if someone would assign `null` to the `Subdirectory` property, or a `null` to the property itself. What behavior would be expected in that case?

For the MVVM aspect of your question, you might want to keep separate your Model from your ViewModel. What I suggest is to have a separate ViewModel class that would expose all the details necessary for the binding.

Design is a complicated topic and it's better if you learn it from experts. I usually recommend to my colleagues [Adaptive Code via C#](https://www.amazon.com/Adaptive-Code-via-principles-Developer/dp/0735683204) and [Encapsulation and SOLID (Video Course)](https://app.pluralsight.com/library/courses/encapsulation-solid/table-of-contents) as a starting point.
